src/v2/stack_aggregator.py

```diff
@@ -0,0 +1,368 @@
+        bindings['packages'] = [pkg.dict(exclude={'dependencies'}) for pkg in pkgs]
+        result = post_gremlin(query, bindings)
+        if result:
+            pkgs_with_vuln['result']['data'] += result['result']['data']
+
+    logger.info(
+        'get_package_details_with_vulnerabilities time: %f total_results %d',
+        time.time() - time_start, len(pkgs_with_vuln['result']['data']))
+    return pkgs_with_vuln['result']['data']
+
+
+def get_package_details_map(
+        graph_response: List[Dict[str, object]]) -> Dict[Package, PackageDetails]:
+    """Transform graph response into PackageDetails map."""
+    package_details: List[Tuple[Package, PackageDetails]] = []
+    for pkg in graph_response:
+        package_details.append(create_package_details(pkg))
+    # covert list of (pkg, package_details) into map
+    return dict(package_details)
+
+
+def _has_vulnerability(pkg: PackageDetails) -> bool:
+    return pkg and (pkg.public_vulnerabilities or pkg.private_vulnerabilities)
+
+
+def get_denormalized_package_details(
+        packages: NormalizedPackages,
+        package_details_map: Dict[Package, PackageDetails]) -> List[PackageDetails]:
+    """Pack PackageDetails according to it's dependency graph structure."""
+    package_details = []
+    for package, transitives in packages.dependency_graph.items():
+        package_detail = package_details_map.get(package)
+        if package_detail:
+            package_detail = package_detail.copy()
+        else:
+            continue  # pragma: no cover
+        transitive_details = []
+        for transitive in transitives:
+            transitive_detail = package_details_map.get(transitive)
+            if _has_vulnerability(transitive_detail):
+                transitive_detail = transitive_detail.copy()
+            else:
+                continue  # pragma: no cover
+            transitive_details.append(transitive_detail)
+        package_detail.dependencies = list(transitives)
+        package_detail.vulnerable_dependencies = transitive_details
+        package_details.append(package_detail)
+    return package_details
+
+
+def get_package_details_from_graph(packages: NormalizedPackages) -> Dict[Package, PackageDetails]:
+    """Get dependency data from graph."""
+    graph_response = get_package_details_with_vulnerabilities(packages)
+    return get_package_details_map(graph_response)
+
+
+def initiate_unknown_package_ingestion(output: StackAggregatorResult):
+    """Ingestion of Unknown dependencies."""
+    try:
+        for dep in output.unknown_dependencies:
+            server_create_analysis(output.ecosystem, dep.name, dep.version, api_flow=True,
+                                   force=False, force_graph_sync=True)
+    except Exception as e:  # pylint:disable=W0703,C0103
+        logger.error('Ingestion has been failed for {%s, %s, %s}',
+                     output.ecosystem, dep.name, dep.version)
+        logger.error(e)
+
+
+class StackAggregator:
```

Comment:
Why do we need a class if everything is static. Also why are some methods outside the class?

Reply:
Yes, I'm working on it.